lib/srv/db/redis/cmds.go

```diff
@@ -0,0 +1,209 @@
+func (e *Engine) processCmd(ctx context.Context, redisClient redis.UniversalClient, cmd *redis.Cmd) error {
+	switch strings.ToLower(cmd.Name()) {
+	case helloCmd:
+		return trace.NotImplemented("RESP3 is not supported")
+	case authCmd:
+		if err := e.processAuth(ctx, redisClient, cmd); err != nil {
+			return trace.Wrap(err)
+		}
+		return nil
+	case subscribeCmd:
+		e.Audit.OnQuery(e.Context, e.sessionCtx, common.Query{Query: cmd.String()})
+		if err := e.subscribeCmd(ctx, redisClient.Subscribe, cmd); err != nil {
+			return trace.Wrap(err)
+		}
+		return nil
+	case psubscribeCmd:
+		e.Audit.OnQuery(e.Context, e.sessionCtx, common.Query{Query: cmd.String()})
+		if err := e.subscribeCmd(ctx, redisClient.PSubscribe, cmd); err != nil {
+			return trace.Wrap(err)
+		}
+		return nil
+	case punsubscribeCmd:
+		// TODO(jakub): go-redis doesn't expose any API for this command. Investigate alternative options.
+		return trace.NotImplemented("PUNSUBSCRIBE is not supported by Teleport")
+	case ssubscribeCmd, sunsubscribeCmd:
+		// go-redis doesn't support Redis 7+ commands yet.
+		return trace.NotImplemented("Redis 7.0+ is not yet supported")
+	default:
+		e.Audit.OnQuery(e.Context, e.sessionCtx, common.Query{Query: cmd.String()})
+
+		// Here the command is sent to the DB.
+		return redisClient.Process(ctx, cmd)
+	}
+}
+
+type redisSubscribeFn = func(ctx context.Context, channels ...string) *redis.PubSub
+
+// subscribeCmd handles subscription in Redis. A new connection is created and maintained to read published messages
+// from Redis and send them back to the client.
+func (e *Engine) subscribeCmd(ctx context.Context, subscribeFn redisSubscribeFn, cmd *redis.Cmd) error {
+	if len(cmd.Args()) < 2 {
+		return redis.RedisError("invalid command")
+	}
+
+	args := make([]string, 0, len(cmd.Args()))
+	for _, a := range cmd.Args()[1:] {
+		args = append(args, a.(string))
+	}
+	pubSub := subscribeFn(ctx, args...)
+	defer func() {
+		if err := pubSub.Close(); err != nil {
+			e.Log.Errorf("failed to close Redis Pub/Sub connection: %v", err)
+		}
+	}()
+
+	return e.processPubSub(ctx, pubSub)
+}
+
+// processPubSub reads messages from Redis and forward them to connected client.
+func (e *Engine) processPubSub(ctx context.Context, pubSub *redis.PubSub) error {
+	for {
+		msg, err := pubSub.Receive(ctx)
+		if err != nil {
+			return err
+		}
+
+		switch msg := msg.(type) {
+		case *redis.Subscription:
+			if err := e.sendToClient([]interface{}{msg.Kind, msg.Channel, msg.Count}); err != nil {
+				return err
+			}
+		case *redis.Pong:
+			if err := e.sendToClient([]interface{}{msg.Payload}); err != nil {
+				return err
+			}
+		case *redis.Message:
+			var payloadResp interface{}
+			if len(msg.PayloadSlice) > 0 {
+				payloadResp = []interface{}{"message", msg.Channel, msg.PayloadSlice}
+			} else {
+				payloadResp = []interface{}{"message", msg.Channel, msg.Payload}
+			}
```

Comment:
Good catch. I was sending back wrong message type if client used `psubscribe` instead of `subscribe`. I've fixed the code and added test to cover that.